requirements.txt

```diff
@@ -7,16 +7,15 @@ matplotlib
 #qml
 # tqdm
 # for documentation
-sphinx>=2.1.2
-breathe>=4.14.1
+sphinx==4.0.1
+breathe==4.30.0
+jinja2==2.11.3
 sphinx_rtd_theme
-nbsphinx>=0.8.1
-pygments>=2.4.1
-# Weird bug in latest version (0.17) of docutils; can remove once fixed
-docutils==0.16
-# Developers tools - code formatting
-cpplint
-black==20.8b1 
+nbsphinx==0.8.5
+pygments==2.9.0
+docutils==0.17
+cpplint==1.5.5
+black==20.8b1
```

Comment:
I'm not fond of pinning all dependencies, but at least these are only development dependencies. 

We will have to be careful to update these regularly to not fall too much behind upstream changes.

Reply:
> I'm just not very fond of all this random upstream breakage in the CI that seems to be happening recently.

That's an issue of pip/the Python ecosystem not having lockfiles and a good dependency update story. I mean, there is no real "per-project dependency" mechanism except duplicating a full python environment and changing your PATH (i.e. virtualenv). So pinning in "requirements.txt" is fine for me, but we need to actively update.